src/common/health.rs

```diff
@@ -0,0 +1,359 @@
+use std::cmp::min;
+use std::collections::HashSet;
+use std::future::{self, Future};
+use std::sync::atomic::{self, AtomicBool};
+use std::sync::Arc;
+use std::time::Duration;
+
+use api::grpc::qdrant::qdrant_internal_client::QdrantInternalClient;
+use api::grpc::qdrant::{GetCommitIndexRequest, GetCommitIndexResponse};
+use api::grpc::transport_channel_pool::{self, TransportChannelPool};
+use collection::shards::replica_set::ReplicaState;
+use collection::shards::shard::ShardId;
+use collection::shards::CollectionId;
+use common::defaults::CONSENSUS_META_OP_WAIT;
+use futures::stream::FuturesUnordered;
+use futures::{StreamExt as _, TryStreamExt as _};
+use storage::content_manager::consensus_manager::ConsensusStateRef;
+use storage::content_manager::toc::TableOfContent;
+use tokio::sync::broadcast;
+use tokio::{runtime, sync, task, time};
+
+pub struct Ready {
+    toc: Arc<TableOfContent>,
+    consensus_state: ConsensusStateRef,
+    runtime: runtime::Handle,
+    is_ready: Arc<AtomicBool>,
+    notify_ready: Arc<sync::Notify>,
+    task: sync::RwLock<Option<task::JoinHandle<()>>>,
+    check_ready: sync::broadcast::Sender<()>,
+}
+
+impl Ready {
+    pub fn new(
+        toc: Arc<TableOfContent>,
+        consensus_state: ConsensusStateRef,
+        runtime: runtime::Handle,
+    ) -> Self {
+        let (check_ready, _) = broadcast::channel(1);
+
+        Self {
+            toc,
+            consensus_state,
+            runtime,
+            is_ready: Default::default(),
+            notify_ready: Default::default(),
+            task: Default::default(),
+            check_ready,
+        }
+    }
+
+    pub async fn check_ready(&self) -> bool {
+        if self.is_ready() {
+            true
+        } else {
+            self.notify_task().await;
+            self.wait_ready().await
+        }
+    }
+
+    pub fn is_ready(&self) -> bool {
+        self.is_ready.load(atomic::Ordering::Relaxed)
+    }
+
+    async fn wait_ready(&self) -> bool {
+        time::timeout(Duration::from_millis(500), self.notify_ready.notified())
```

Comment:
This can data race. Even if ready now, we aren't guaranteed to receive a notification here when this is called from `check_ready()`.

An example scenario would be that we become ready right after the `is_ready()` check, after which `notify_task()` returns early and `wait_ready()` doesn't get notified.

It wouldn't be a disaster, but nice to fix anyway.

Reply:
Oh, you're right.

It *won't* data race, if we use `Notify::notify_one` (instead of `Notify::notify_waiters`) in `Task::exec` to prevent data race. Didn't read the docs for `Notify::notify_waiters` carefully enough...

Reply:
Still may be possible for subscriber to miss both initial `self.is_ready()` and `self.notify_ready.notified()`, but it will resolve correctly after 500ms wait.

Reply:
Or what @generall said, it's just a bit less convenient to use.

Reply:
Fixed as @generall suggested.